syft/execution/protocol.py

```diff
@@ -1,479 +1,560 @@
-        """
-        if self.workers_resolved:
-            raise RuntimeError(
-                f"This protocol is already deployed to {', '.join(worker.id for worker, plan in self.plans)}."
-            )
+        if not hasattr(self, "forward"):
+            self.forward = forward_func or None
 
-        n_workers = len(set(worker for worker, plan in self.plans))
+        self.__name__ = self.__repr__()  # For PyTorch jit tracing compatibility
 
-        # to correctly map workers, we must have exactly 1 worker for 1 plan
-        # or 1 worker for 1 fictive worker.
-        # If we don't, raise here
-        if len(self.plans) != len(workers) != n_workers:
-            raise RuntimeError(
-                f"This protocol is designed for {n_workers} workers, but {len(workers)} were provided."
-            )
+        # List of available translations
+        # self.translations = []
 
-        self._resolve_workers(workers)
+    # @property
+    # def state(self):
+    #     return self.role.state
 
-        # update plans list with *pointers* to the plans
-        self.plans = [(worker, plan.send(worker)) for (worker, plan) in self.plans]
+    # def parameters(self):
+    #     """
+    #     This is defined to match the torch api of nn.Module where .parameters() return the model tensors / parameters
+    #     """
+    #     if self.state is not None:
+    #         return self.state.tensors()
+    #     else:
+    #         return []
 
-        return self
+    def get_role_for_owner(self, owner):
+        if owner.id not in self.roles:
+            self.roles[owner.id] = Role()
+        return self.roles[owner.id]
+
+    def build(self, *args):
+        """Builds the protocol.
 
-    def __call__(self, *args, **kwargs):
-        return self.run(*args, **kwargs)
+        First, run the function to be converted in a protocol in a context which
+        activates the tracing and record the actions in trace.logs
+
+        Second, store the result ids temporarily to helper ordering the output
+        placeholders at return time
+
+        Third, loop through the trace logs and replace the tensors found in the
+        actions logged by PlaceHolders. Record those actions in
+        protocol.actions
+
+        Args:
+            args: Input arguments to run the protocol
+        """
+        # Reset previous build
+        self.roles = {}
+
+        # Enable tracing
+        self.toggle_tracing(True)
+        self.is_building = True
+
+        # Run once to build the protocol
+        ph_args = tuple()
+        for arg in args:
+            arg_role = self.get_role_for_owner(arg.owner)
+
+            ph_arg = PlaceHolder.create_from(arg, owner=arg.owner, role=arg_role, tracing=True)
+            # Register inputs in role
+            arg_role.register_input(ph_arg)
+
+            ph_args += (ph_arg,)
+
+        # Add state to args if needed
+        # if self.include_state:
+        #     ph_args += (self.state,)
```

Comment:
Having read down the PR to this point, starting to think we should rip the concept of `State` out of `Protocol` for now. We can leave it in `Role` of course, so that `Plan` keeps working, but it seems like it's only adding complications that we're not ready to deal with here.

Reply:
I think states will be important in real cases but yeah, as the code is in Plan anyway, I can just remove that totally